extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/output/S3UploadConfig.java

```diff
@@ -0,0 +1,115 @@
+ */
+@LazySingleton
+public class S3UploadConfig
+{
+  /**
+   * The maximum chunk size based on injected values for {@link S3OutputConfig} and {@link S3ExportConfig} used for computing maximum number of chunks to save on disk at any given point in time.
+   * It is initialized to 5 MiB which is the minimum chunk size possible, denoted by {@link S3OutputConfig#S3_MULTIPART_UPLOAD_MIN_PART_SIZE_BYTES}.
+   */
+  private long chunkSize = new HumanReadableBytes("5MiB").getBytes();
+
+  /**
+   * An atomic counter to track the current number of chunks saved to local disk.
+   */
+  private final AtomicInteger currentNumChunks = new AtomicInteger(0);
+
+  /**
+   * The maximum number of chunks that can be saved to local disk concurrently.
+   * This value is recalculated when the chunk size is updated in {@link #updateChunkSizeIfGreater(long)}.
+   */
+  private int maxConcurrentNumChunks = 100;
+
+  private static final Logger log = new Logger(S3UploadConfig.class);
+
+  /**
+   * Increments the counter for the current number of chunks saved on disk.
+   */
+  public void incrementCurrentNumChunks()
+  {
+    currentNumChunks.incrementAndGet();
+  }
+
+  /**
+   * Decrements the counter for the current number of chunks saved on disk.
+   */
+  public void decrementCurrentNumChunks()
+  {
+    currentNumChunks.decrementAndGet();
+  }
+
+  /**
+   * Gets the current number of chunks saved to local disk.
+   *
+   * @return the current number of chunks saved to local disk.
+   */
+  public int getCurrentNumChunks()
+  {
+    return currentNumChunks.get();
+  }
+
+  /**
+   * Gets the maximum number of concurrent chunks that can be saved to local disk.
+   *
+   * @return the maximum number of concurrent chunks that can be saved to local disk.
+   */
+  public int getMaxConcurrentNumChunks()
+  {
+    return maxConcurrentNumChunks;
+  }
+
+  /**
+   * Updates the chunk size if the provided size is greater than the current chunk size.
+   * Recomputes the maximum number of concurrent chunks that can be saved to local disk based on the new chunk size.
+   *
+   * @param chunkSize the new chunk size to be set if it is greater than the current chunk size.
+   */
+  public void updateChunkSizeIfGreater(long chunkSize)
+  {
+    if (this.chunkSize < chunkSize) {
+      this.chunkSize = chunkSize;
+      recomputeMaxConcurrentNumChunks();
+    }
+  }
+
+  /**
+   * Recomputes the maximum number of concurrent chunks that can be saved to local disk based on the current chunk size.
+   * The maximum allowed chunk size is {@link S3OutputConfig#S3_MULTIPART_UPLOAD_MAX_PART_SIZE_BYTES} which is quite big,
+   * so we restrict the maximum disk space used to the same, at any given point in time.
+   */
+  private void recomputeMaxConcurrentNumChunks()
+  {
+    maxConcurrentNumChunks = (int) (S3OutputConfig.S3_MULTIPART_UPLOAD_MAX_PART_SIZE_BYTES / chunkSize);
+    log.info("Recomputed maxConcurrentNumChunks: %d", maxConcurrentNumChunks);
```

Comment:
How often do we expect this to be called?

Reply:
Twice. Please see: https://github.com/apache/druid/pull/16481#discussion_r1616620124